godot-core/src/task/async_runtime.rs

```diff
@@ -0,0 +1,456 @@
+    fn clear_task(&mut self, index: usize) {
+        self.tasks[index].clear();
+    }
+
+    /// Move a future back into its slot.
+    ///
+    /// # Panic
+    /// - If the underlying slot is not in the [`FutureSlotState::Polling`] state.
+    fn park_task(&mut self, index: usize, future: Pin<Box<dyn Future<Output = ()>>>) {
+        self.tasks[index].park(future);
+    }
+
+    /// Track that a future caused a panic.
+    ///
+    /// This is only available for itest.
+    #[cfg(feature = "trace")]
+    fn track_panic(&mut self, task_id: u64) {
+        self.panicked_tasks.insert(task_id);
+    }
+}
+
+trait WithRuntime {
+    fn with_runtime<R>(&'static self, f: impl FnOnce(&AsyncRuntime) -> R) -> R;
+    fn with_runtime_mut<R>(&'static self, f: impl FnOnce(&mut AsyncRuntime) -> R) -> R;
+}
+
+impl WithRuntime for LocalKey<RefCell<Option<AsyncRuntime>>> {
+    fn with_runtime<R>(&'static self, f: impl FnOnce(&AsyncRuntime) -> R) -> R {
+        self.with_borrow(|rt| {
+            let rt_ref = rt.as_ref().expect(ASYNC_RUNTIME_DEINIT_PANIC_MESSAGE);
+
+            f(rt_ref)
+        })
+    }
+
+    fn with_runtime_mut<R>(&'static self, f: impl FnOnce(&mut AsyncRuntime) -> R) -> R {
+        self.with_borrow_mut(|rt| {
+            let rt_ref = rt.as_mut().expect(ASYNC_RUNTIME_DEINIT_PANIC_MESSAGE);
+
+            f(rt_ref)
+        })
+    }
+}
+
+/// Use a godot waker to poll it's associated future.
+///
+/// # Panics
+/// - If called from a thread other than the main-thread.
+fn poll_future(godot_waker: Arc<GodotWaker>) {
+    let current_thread = thread::current().id();
+
+    assert_eq!(
+        godot_waker.thread_id,
+        current_thread,
+        "trying to poll future on a different thread!\n  Current thread: {:?}\n  Future thread: {:?}",
+        current_thread,
+        godot_waker.thread_id,
+    );
+
+    let waker = Waker::from(godot_waker.clone());
+    let mut ctx = Context::from_waker(&waker);
+
+    // Move future out of the runtime while we are polling it to avoid holding a mutable reference for the entire runtime.
+    let future = ASYNC_RUNTIME.with_runtime_mut(|rt| {
+        match rt.take_task_for_polling(godot_waker.runtime_index, godot_waker.task_id) {
+            FutureSlotState::Empty => {
+                panic!("Future slot is empty when waking it! This is a bug!");
+            }
+
+            FutureSlotState::Gone => None,
+
+            FutureSlotState::Polling => {
+                panic!("The same GodotWaker has been called recursively, this is not expected!");
+            }
+
+            FutureSlotState::Pending(future) => Some(future),
+        }
+    });
+
+    let Some(future) = future else {
+        // Future has been canceled while the waker was already triggered.
+        return;
+    };
+
+    let error_context = || "Godot async task failed";
+
+    // Future should be unwind-safe as we will immediately drop it in cases where a panic occurs.
+    let mut future = AssertUnwindSafe(future);
```

Comment:
```suggestion
    // If Future::poll() panics, the future is immediately dropped and cannot be accessed again,
    // thus any state that may not have been unwind-safe cannot be observed later.
    let mut future = AssertUnwindSafe(future);
```

(How I interpreted our earlier discussion in https://github.com/godot-rust/gdext/pull/1043#discussion_r1976712715)